sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -122,46 +119,48 @@ case class CostBasedJoinReorder(conf: CatalystConf) extends Rule[LogicalPlan] wi
  * level 3: p({A, B, C, D})
  * where p({A, B, C, D}) is the final output plan.
  *
- * For cost evaluation, since physical costs for operators are not available currently, we use
- * cardinalities and sizes to compute costs.
+ * To evaluate cost for a given plan, we calculate the sum of cardinalities for all intermediate
+ * joins in the plan.
  */
 object JoinReorderDP extends PredicateHelper {
 
   def search(
       conf: CatalystConf,
       items: Seq[LogicalPlan],
-      conditions: Set[Expression],
-      topOutput: AttributeSet): Option[LogicalPlan] = {
+      conditions: Set[Expression]): Option[LogicalPlan] = {
 
     // Level i maintains all found plans for i + 1 items.
     // Create the initial plans: each plan is a single item with zero cost.
-    val itemIndex = items.zipWithIndex
+    val itemIndex = items.zipWithIndex.map(_.swap).toMap
     val foundPlans = mutable.Buffer[JoinPlanMap](itemIndex.map {
-      case (item, id) => Set(id) -> JoinPlan(Set(id), item, Set(), Cost(0, 0))
-    }.toMap)
+      case (id, item) => Set(id) -> JoinPlan(Set(id), item, cost = 0)
+    })
 
-    for (lev <- 1 until items.length) {
+    while (foundPlans.size < items.length && foundPlans.last.size > 1) {
```

Comment:
add some comments to explain why we can stop when the last level has less than 1 entry.